apps/client/src/features/app-settings/panel/automations-panel/BlueprintForm.tsx

```diff
@@ -0,0 +1,484 @@
+import { IoAdd } from '@react-icons/all-files/io5/IoAdd';
+import { IoTrash } from '@react-icons/all-files/io5/IoTrash';
+import {
+  AutomationBlueprint,
+  AutomationBlueprintDTO,
+  CustomFields,
+  HTTPOutput,
+  isHTTPOutput,
+  isOSCOutput,
+  OntimeEvent,
+  OSCOutput,
+} from 'ontime-types';
+
+import { addBlueprint, editBlueprint, testOutput } from '../../../../common/api/automation';
+import { maybeAxiosError } from '../../../../common/api/utils';
+import Tag from '../../../../common/components/tag/Tag';
+import useAutomationSettings from '../../../../common/hooks-query/useAutomationSettings';
+import useCustomFields from '../../../../common/hooks-query/useCustomFields';
+import { preventEscape } from '../../../../common/utils/keyEvent';
+import { isIPAddress, startsWithHttp } from '../../../../common/utils/regex';
+import * as Panel from '../../panel-utils/PanelUtils';
+
+import style from './BlueprintForm.module.scss';
+
+interface BlueprintFormProps {
+  blueprint: AutomationBlueprintDTO | AutomationBlueprint;
+  onClose: () => void;
+}
+
+export default function BlueprintForm(props: BlueprintFormProps) {
+  const { blueprint, onClose } = props;
+  const isEdit = isBlueprint(blueprint);
+  const { data } = useCustomFields();
+  const { refetch } = useAutomationSettings();
+  const fieldList = useMemo(() => makeFieldList(data), [data]);
+
+  const {
+    control,
+    handleSubmit,
+    getValues,
+    register,
+    setError,
+    setFocus,
+    formState: { errors, isSubmitting, isDirty, isValid },
+  } = useForm<AutomationBlueprintDTO>({
+    mode: 'onChange',
+    defaultValues: {
+      title: blueprint?.title ?? '',
+      filterRule: blueprint?.filterRule ?? 'all',
+      filters: blueprint?.filters ?? [],
+      outputs: blueprint?.outputs ?? [],
+    },
+    resetOptions: {
+      keepDirtyValues: true,
+    },
+  });
+
+  const {
+    fields: fieldFilters,
+    append: appendFilter,
+    remove: removeFilter,
+  } = useFieldArray({
+    name: 'filters',
+    control,
+  });
+
+  const {
+    fields: fieldOutputs,
+    append: appendOutput,
+    remove: removeOutput,
+  } = useFieldArray({
+    name: 'outputs',
+    control,
+  });
+
+  // give initial focus to the title field
+  useEffect(() => {
+    setFocus('title');
+  }, [setFocus]);
+
+  const handleAddNewFilter = () => {
+    appendFilter({ field: '', operator: 'equals', value: '' });
+  };
+
+  const handleAddNewOSCOutput = () => {
+    // @ts-expect-error -- we dont want to pass a port to the new object
+    appendOutput({ type: 'osc', targetIP: '', targetPort: undefined, address: '', args: '' });
+  };
+
+  const handleAddNewHTTPOutput = () => {
+    appendOutput({ type: 'http', url: '' });
+  };
+
+  const handleTestOSCOutput = async (index: number) => {
+    try {
+      const values = getValues(`outputs.${index}`) as OSCOutput;
+      if (!values.targetIP || !values.targetPort || !values.address || !values.args) {
```

Comment:
This if statement currently prevents testing OSC output if no args are specified.